lib/Optimizer/Transforms/ApplyOpSpecialization.cpp

```diff
@@ -337,40 +376,88 @@ class ApplySpecializationPass
     }
   }
 
-  func::FuncOp createControlVariantOf(func::FuncOp funcOp) {
+  /// Look for quake.compute_action operations or quake.apply triple patterns in
+  /// the FuncOp \p func. In these cases, we do not want to add the controls to
+  /// the compute and uncompute functions.
+  DenseSet<Operation *> computeActionAnalysis(func::FuncOp func) {
+    DenseSet<Operation *> controlNotNeeded;
+    func->walk([&](Operation *op) {
+      if (auto compAct = dyn_cast<quake::ComputeActionOp>(op)) {
+        // This is clearly a compute action. Mark the compute side.
+        if (auto *defOp = compAct.getCompute().getDefiningOp()) {
+          controlNotNeeded.insert(defOp);
+        } else {
+          compAct.emitError("compute value not determined");
+          signalPassFailure();
+        }
+      } else if (auto app0 = dyn_cast<quake::ApplyOp>(op)) {
+        auto next1 = ++app0->getIterator();
+        Operation &op1 = *next1;
+        if (auto app1 = dyn_cast<quake::ApplyOp>(op1)) {
+          auto next2 = ++next1;
+          Operation &op2 = *next2;
+          if (auto app2 = dyn_cast<quake::ApplyOp>(op2);
+              app2 && (app0.getCalleeAttr() == app2.getCalleeAttr()) &&
+              ((!app0.getIsAdj() && app2.getIsAdj()) ||
+               (app0.getIsAdj() && !app2.getIsAdj())) &&
+              !controlNotNeeded.count(app1)) {
+            // This is a compute_action lowered to 3 successive apply
+            // operations. We want to add the control to ONLY the action, the
+            // middle apply op, so mark the compute and uncompute applies.
+            controlNotNeeded.insert(app0);
+            controlNotNeeded.insert(app2);
+          }
+        }
+      }
+    });
+    return controlNotNeeded;
+  }
+
+  func::FuncOp createControlVariantOf(func::FuncOp func) {
     ModuleOp module = getOperation();
     auto *ctx = module.getContext();
-    auto funcName = getCtrlVariantFunctionName(funcOp.getName().str());
-    auto funcTy = funcOp.getFunctionType();
+    // Perform a pre-analysis to determine if func has any compute_action like
+    // ops. If it does, then there is an exception case. Instead of applying the
+    // controls to the compute kernel, just use the compute kernel (and
+    // uncompute kernel) without the controls added.
+    auto funcName = getCtrlVariantFunctionName(func.getName().str());
+    auto funcTy = func.getFunctionType();
     auto qvecTy = quake::QVecType::getUnsized(ctx);
-    auto loc = funcOp.getLoc();
+    auto loc = func.getLoc();
     SmallVector<Type> inTys = {qvecTy};
     inTys.append(funcTy.getInputs().begin(), funcTy.getInputs().end());
     auto newFunc = cudaq::opt::factory::createFunction(
         funcName, funcTy.getResults(), inTys, module);
     newFunc.setPrivate();
     IRMapping mapping;
-    funcOp.getBody().cloneInto(&newFunc.getBody(), mapping);
+    func.getBody().cloneInto(&newFunc.getBody(), mapping);
+    auto controlNotNeeded = computeActionAnalysis(newFunc);
```

Comment:
Is it possible to setup a flag or something such that we can control whether we do this or not? The flag should always default to true, and it is probably best that it is not something we document or talk too much about (there's no reason real users will need this). My objective here is to provide a way to demonstrate key features about the compiler - plots that show how the code generated here is optimal and leads to a decrease in the number of controlled operations required. We may also want this for testing with CircuitCheck (make sure the unitaries are the same if we do this and if we don't).

Reply:
it's ok to not be "truly" hidden.

Reply:
Done.